spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java

```diff
@@ -326,6 +339,73 @@ public void setSkipSameTopicFatalExceptions(boolean skipSameTopicFatalExceptions
 		this.skipSameTopicFatalExceptions = skipSameTopicFatalExceptions;
 	}
 
+	/**
+	 * Set a {@link ExceptionHeadersCreator} implementation to completely take over
+	 * setting the exception headers in the output record. Disables all headers that are
+	 * set by default.
+	 * @param headersCreator the creator.
+	 * @since 2.8.4
+	 */
+	public void setExceptionHeadersCreator(ExceptionHeadersCreator headersCreator) {
+		Assert.notNull(headersCreator, "'headersCreator' cannot be null");
+		this.exceptionHeadersCreator = headersCreator;
+	}
+
+	/**
+	 * Clear the header inclusion bit for the header name.
+	 * @param headers the headers to clear.
+	 * @since 2.8.4
+	 */
+	public void excludeHeader(HeaderNames.HeadersToAdd... headers) {
+		Assert.notNull(headers, "'headers' cannot be null");
+		Assert.noNullElements(headers, "'headers' cannot include null elements");
+		for (HeaderNames.HeadersToAdd header : headers) {
+			this.whichHeaders.remove(header);
+		}
+	}
+
+	/**
+	 * Set the header inclusion bit for the header name.
+	 * @param headers the headers to set.
+	 * @since 2.8.4
+	 */
+	public void includeHeader(HeaderNames.HeadersToAdd... headers) {
+		Assert.notNull(headers, "'headers' cannot be null");
+		Assert.noNullElements(headers, "'headers' cannot include null elements");
+		for (HeaderNames.HeadersToAdd header : headers) {
+			this.whichHeaders.add(header);
+		}
+	}
+
+	/**
+	 * Add a function which will be called to obtain additional headers to add to the
+	 * published record. Functions are called in the order that they are added, and after
+	 * any function passed into {@link #setHeadersFunction(BiFunction)}.
+	 * @param headersFunction the headers function.
+	 * @since 2.8.4
+	 * @see #setHeadersFunction(BiFunction)
+	 */
+	public void addHeadersFunction(BiFunction<ConsumerRecord<?, ?>, Exception, Headers> headersFunction) {
+		Assert.notNull(headersFunction, "'headersFunction' cannot be null");
+		if (this.headersFunction.equals(DEFAULT_HEADERS_FUNCTION)) {
+			this.headersFunction = headersFunction;
+		}
+		else {
+			BiFunction<ConsumerRecord<?, ?>, Exception, Headers> toCompose = this.headersFunction;
+			this.headersFunction = (rec, ex) -> {
+				Headers headers1 = toCompose.apply(rec, ex);
+				if (headers1 == null) {
+					headers1 = new RecordHeaders();
+				}
+				Headers headers2 = headersFunction.apply(rec, ex);
+				if (headers2 != null) {
+					headers2.forEach(headers1::add);
+				}
```

Comment:
There's one minor point of concern with this implementation. Imagine that the original `toCompose` function does something similar to this:

```java
Headers addHeaders(ConsumerRecord<...> record, Exception exception) {
    RecordHeaders extraHeaders = new RecordHeaders();
    extraHeaders.add("key", ...);
    extraHeaders.setReadOnly(); // This is the important line.

    return extraHeaders;
}
```

The returned value would be stored in `headers1`. `headers2.forEach(headers1::add);` would then fail with an `IllegalStateException` because `headers1` can no longer be modified. A safer, albeit slightly more expensive, alternative could be:

```java
// I believe the constructor is null-safe, but please double-check.
RecordHeaders headers1 = new RecordHeaders(toCompose.apply(rec, ex));
Headers headers2 = headersFunction.apply(rec, ex);
if (headers2 != null) {
    headers2.forEach(headers1::add);
}
return headers1;
```

Reply:
Unusual; but thanks.

Reply:
Yeah, I highly doubt anyone would actually do it, so I'm happy with the current implementation too.